src/components/Transaction/Form.tsx

```diff
@@ -0,0 +1,199 @@
+import {
+  FILECOIN_NUMBER_PROPTYPE,
+  MESSAGE_PROPTYPE,
+  TxState,
+  TX_STATE_PROPTYPE
+} from '../../customPropTypes'
+import { MessagePending } from '../../generated/graphql'
+import { TransactionButtons } from './Buttons'
+import { TransactionFee } from './Fee'
+import { TransactionHeader } from './Header'
+import { TransactionTotal } from './Total'
+import { Dialog, ShadowBox } from '../Layout'
+import { useSubmittedMessages } from '../HistoryTables/PendingMsgContext'
+import { getMaxGasFee } from '../../utils'
+import { useWallet, useWalletProvider } from '../../services'
+import { logger } from '../../logger'
+
+export const TransactionForm = ({
+  children,
+  title,
+  description,
+  message,
+  total,
+  txState,
+  setTxState,
+  maxFee,
+  txFee,
+  setTxFee,
+  getParams,
+  onComplete
+}: TransactionFormProps) => {
+  const router = useRouter()
+  const wallet = useWallet()
+  const { pushPendingMessage } = useSubmittedMessages()
+  const { loginOption, walletError, getProvider } = useWalletProvider()
+
+  // Transaction states
+  const [txError, setTxError] = useState<Error | null>(null)
+  const [gasParamsError, setGasParamsError] = useState<Error | null>(null)
+  const [messageWithGas, setMessageWithGas] = useState<Message | null>(null)
+
+  // Max transaction fee used for getting gas params. Will be
+  // null until the user manually changes the transaction fee.
+  const [inputFee, setInputFee] = useState<FilecoinNumber | null>(null)
+
+  // Calculate transaction fee (fee cap times limit)
+  useEffect(() => {
+    if (!messageWithGas) return setTxFee(null)
+    const feeCapStr = messageWithGas.gasFeeCap.toFixed(0, BigNumber.ROUND_CEIL)
+    const feeCap = new FilecoinNumber(feeCapStr, 'attofil')
+    const limit = messageWithGas.gasLimit
+    setTxFee(getMaxGasFee(feeCap, limit))
+  }, [setTxFee, messageWithGas])
+
+  //Load gas params with the current message and input fee
+  const getGasParams = async () => {
+    setTxState(TxState.LoadingTxFee)
+    setGasParamsError(null)
+    setMessageWithGas(null)
+    try {
+      const provider = await getProvider()
+      setMessageWithGas(
+        await provider.gasEstimateMessageGas(
+          message.toLotusType(),
+          inputFee ? inputFee.toAttoFil() : undefined
+        )
+      )
+    } catch (e: any) {
+      logger.error(e)
+      setGasParamsError(e)
+    } finally {
+      setTxState(TxState.FillingTxFee)
+    }
+  }
+
+  // Attempt sending message
+  const onSend = async () => {
+    setTxState(TxState.LoadingTxDetails)
+    setTxError(null)
+    try {
+      const provider = await getProvider()
+      const nonce = await provider.getNonce(wallet.address)
+      const newMessage = new Message({
+        to: messageWithGas.to,
+        from: messageWithGas.from,
+        nonce,
+        value: messageWithGas.value,
+        method: messageWithGas.method,
+        params: getParams ? getParams(nonce) : messageWithGas.params,
```

Comment:
Hmm yeah I implemented this for the safe.

So for getting the gas parameters, we first create the base64 params with a nonce of `0`, as the nonce is encoded in the parameters. When we actually send the message, we get the nonce and need to recreate the base64 params with the actual nonce. This component can't be aware of how the params are created because it's different for each msig method.

The only way we can remove this prop I think is to get the nonce in the parent component already and pass it to the Form component. The caveat is that the nonce might be invalid if multiple tx windows were opened at the same time.